packages/astro/src/runtime/server/render/page.ts

```diff
@@ -43,8 +44,9 @@ export async function renderPage(
 
 	// Mark if this page component contains a <head> within its tree. If it does
 	// We avoid implicit head injection entirely.
+	const propagationHint = getPropagationHint(result, componentFactory);
 	result._metadata.headInTree =
-		result.componentMetadata.get(componentFactory.moduleId!)?.containsHead ?? false;
+		result.componentMetadata.get(componentFactory.moduleId!)?.containsHead ?? propagationHint === 'in-tree';
```

Comment:
should this also be checking for propagationHint === 'self'?

Reply:
Actually, thinking about it again, the `headInTree` variable is intended to be used to know if the `<head>` element is anywhere in the tree of components. That is the actual `<head>` tag. It's not intended to be used to know if we should do head propagation.

The purpose is to support cases where there is no explicit `<head>` element used in the page (a legacy mistake, imo).